README.md
Netuitive Cookbook (Chef)
==========================

[![Build Status](https://travis-ci.org/Netuitive/chef-netuitive.svg?branch=master)](https://travis-ci.org/Netuitive/chef-netuitive) [![Join the chat at https://gitter.im/Netuitive/chef-netuitive](https://badges.gitter.im/Netuitive/chef-netuitive.svg)](https://gitter.im/Netuitive/chef-netuitive?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) [![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/Netuitive/chef-netuitive/master/LICENSE)

A cookbook to automate the installation and configuration of the Netuitive Linux agent. For more
information on the Netuitive Linux Agent, see the [help docs](https://help.netuitive.com/Content/Misc/Datasources/Netuitive/new_netuitive_datasource.htm) or contact Netuitive support at [support@netuitive.com](mailto:support@netuitive.com).

This cookbook is meant to be consumed by wrapper cookbooks such as this: [wrapper cookbook](https://github.com/CloudCruiser/ops_chef-cc_netuitive)

### Attributes

| Key | Type | Description | Default |
|-----|------|-------------|---------|
| `node['netuitive']['version']` | String | The version of the agent to install | `'0.2.9-98'`|
| `node['netuitive']['repo']['urls']` | Hash | A hash of platform specific repo urls | `{ 'debian' => 'https://repos.app.netuitive.com/deb/', 'rhel' => 'https://repos.app.netuitive.com/rpm/noarch' }` |
| `node['netuitive']['repo']['keys']` | Hash | A hash of platform specific repo gpg key locations | `{ 'debian' => 'https://repos.app.netuitive.com/netuitive.gpg', 'rhel' => 'https://repos.app.netuitive.com/RPM-GPG-KEY-netuitive' }` |
| `node['netuitive']['repo']['components']` | Hash | A hash of platform specific compnents | `{ 'debian' => ['stable', 'main'] }` |
| `node['netuitive']['custom_collectors']` | Hash | A hash of collectors and options | `{}` |

Supported Platforms
--------------------

### Operating Systems

#### Official
Tested platforms by travisci:
- centos 6
- centos 7
- ubuntu 16.04
- ubuntu 18.04

Previously tested locally with vagrant or other at one point:
- Ubuntu 18.04 LTS
- Ubuntu 16.04 LTS
- Ubuntu 14.04 LTS
- CentOS 6.7+
- Centos 7.2+
- Debian 8



#### Unofficial
We will attempt to support as many linux distributions as possible and are hoping to expand the above list over time. Any EPEL based system that still supports yum will likely work and we are open to PRs to expands functionality.

### Chef Versions

The `master` branch is intended to be in line with the latest version of chef and its community based distributions. Whenever a breaking change is required to accomplish this we create a branch named `chef-client/$major_version.x`. Maintainers will only backport into the latest major version to match chefs [EOL Products](https://docs.chef.io/versions.html#end-of-life-eol-products) and their corresponding community distributions. Releases on Github and the Chef Supermarket will only be done from the master branches to support the latest versions of chef. EOL versions will continued to be versioned in their respective branches and can be easily pulled in via [berkshelf](https://docs.chef.io/berkshelf.html) which should support pulling from a branch or from a specific sha1.

#### Chef 14.x

This is currently supported by the master branch.

#### Chef 13.x

We supported chef 13 for quite a while after its EOL but its time has come. Please see `chef-client/13.x` branch for continued usage and head branch for pull request.

#### Chef 12.x

We supported chef 12 for quite a while after its EOL but its time has come. Please see `chef-client/12.x` branch for coninued usage and head branch for pull request.

#### Chef 11.x

This cookbook was originally written using many of the nicer things of chef 12. We are open to community contributions to enable/improve the cookbook to support older versions of chef within reason. If you need chef 11 support see the  `chef-client/11.x` branch.

Using the Netuitive Cookbook
-----------------------------

### Recipes
All recipes are simple wrappers around the lightweight resources and providers (LWRPs). We suggest using LWRPs over recipes as it will provide flexibility.

| Name | Description |
|:------:|-------------|
| netuitive::default | Does nothing. |
| netuitive::add_repo | Adds the Netuitive repo. |
| netuitive::configure | Sets base and custom config. |
| netuitive::install_agent | Installs the agent. |

### LWRPs

#### netuitive_configure

##### Actions
`:create`

##### Attributes
| Name | Description | Default |
|:------:|-------------|-------------|
| api_key | Your datasource's API key. | `'CHANGE_ME_PLZ'` |
| api_url | The API url for netuitive. | `'https://api.app.netuitive.com/ingest/infrastructure'` |
| auto_restart_service_on_change | Whether or not to restart the Metricly agent when the configuration changes. You might disable this if your wrapper cookbook is executed by a service which the Metricly agent depends on directly or indirectly and restarting the Metricly service would result in a cyclic dependency. | `true` |
| batch_size | Number of samples to store before sending to Metricly | `500` |
| conf_path | The path to your Netuitive agent config file. | `'/opt/netuitive-agent/conf/netuitive-agent.conf'` |
| cookbook_template | Specifies a different cookbook that the template can come from. | `'netuitive'` |
| disk_space_collector_exclude_filters | Specifies an exclude filter for metrics | `^/boot, ^/mnt` |
| disk_usage_collector_metrics_whitelist | Specifies the metrics whitelist for the DiskUsageCollector. You might change this if you wanted to ignore the Docker device mapper metrics. | `'(?:^.*\.io$|^.*\.average_queue_length$|^.*\.await$|^.*\.iops$|^.*\.read_await$|^.*\.reads$|^.*\.util_percentage|^.*\.write_await$|^.*\.writes$)'` |
| docker_collector_enabled | Whether or not to enable the Docker collector. May be `true` or `false`. | `false` |
| docker_collector_metrics_whitelist | Specifies the metrics whitelist for the NetuitiveDockerCollector. | `.*` |
| relations | An array of relations. | `[]` |
| source | The name of the template. | `'netuitive-agent.conf.erb'` |
| statsd_enabled | Whether to enable embedded statsd server. Specify the string `'True'` or `'False'` | `'False'` |
| statsd_forward | Whether or not to forward stats from the embedded statsd server. May be `true` or `false` | `false`
| statsd_forward_ip | The IP to forward statsd data to if forwarding is enabled. | `'127.0.0.1'`
| statsd_forward_port | The port to forward statsd data to if forwarding is enabled. | `9125`
| statsd_listen_ip | The interface to listen on if statsd is enabled. | `'127.0.0.1'`
| statsd_listen_port | The port the embedded statsd listens on | `8125` |
| statsd_prefix | Prefix applied to your statsd metrics | `statsd` |
| tags | An array of tags . | `[]` |

#### netuitive_collector

##### Actions
`:create`

##### Attributes
| Name | Description | Default |
|:------:|-------------|-------------|
| auto_restart_service_on_change | Whether or not to restart the Metricly agent when the configuration changes. You might disable this if your wrapper cookbook is executed by a service which the Metricly agent depends on directly or indirectly and restarting the Metricly service would result in a cyclic dependency. | `true` |
| conf_path | The path to your Netuitive agent config file. | `'/opt/netuitive-agent/conf/netuitive-agent.conf'` |
| cookbook_template | Specifies a different cookbook that the template can come from. | `'netuitive'` |
| collectors_dir | Dir that custom collectors live in. | `'/opt/netuitive-agent/conf/collectors'` |
| custom_collectors | A hash of collectors and options to create. | `{}` |
| source | The name of the template. | `'collector_generic.conf.erb'` |

#### netuitive_install

##### Actions
`:install`

##### Attributes
| Name | Description | Default |
|:------:|-------------|-------------|
| package_name | The package's name. | `'netuitive-agent'` |

#### netuitive_repo

##### Actions
`:add`

##### Attributes
| Name | Description | Default |
|:------:|-------------|-------------|
| repo_components | A hash of platform-specific components. | `nil` |
| repo_keys | A hash of platform-specific repository GPG keys. | `nil` |
| repo_priority_pins | A hash of platform-specific repo pins. | `nil` |
| repo_urls | A hash of platform-specific repository URLs. | `nil` |
| use_epel_repos | Bool value to enable EPEL repos (doesn't do anything on Debian-based repos). | `nil` |
| version | The version to pin. | `nil` |


Additional Information
-----------------------

### License and Authors
This software is licensed under MIT license quoted below:

```
The MIT License (MIT)

Copyright (c) 2016 Ben Abrams

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```


CHANGELOG.md
# Change Log
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
- added `auto_restart_service_on_change` option, which defaults to true to keep the existing behavior (@ziggythehamster)

## 1.0.0
- document support for long term maintainability (@majormoses)
- drop `< 14`chef support (@majormoses)
- locked `apt` cookbook dependency to ensure `apt_repo` and `apt_preference` are both coming from chef itself rather than the cookbook as this breaks chef 14 clients (@majormoses)

## 0.22.1
- fix/define `docker_collector_metrics_whitelist` option

## 0.22.0
- added `docker_collector_metrics_whitelist` (@rclagett)

## 0.21.0.2-0.21.6
- no-op testing travis supermarket deploys

## 0.21.1
- properly reflected in metadata the reality regarding licensing and chef_version support (@majormoses)

## 0.21.0
- added `batch_size` @johvet

# 0.20.0
- added `disk_space_collector_exclude_filters` @johvet

# 0.19.0
- added `statsd_prefix` option to `netuitive_configure`

# 0.18.0
- added travis deploys to supermarket @TheConnMan

# 0.17.1
- changed to use chef-client 12.16
- fixed issue with apt repository creation per #49. @Createor

# 0.17.0
- bump default collector agent version @Createor

# 0.16.0
- added support for Ubuntu 16.04 LTS and made it an an officially supported platforms
- updating gem dependencies
- removed ruby 2.1 support
- removed chef 11 testing as its support exists in its own branch and I would not expect any more work done on it.
- setup travis to run kitchen tests for centos-6 and centos-7 (more to come later)

# 0.15.1
- fixes template issues when using statsd that were caused in #46. @ziggythehamster

# 0.15.0
- added `statsd_forward`, `statsd_forward_ip`, `statsd_forward_port`, `statsd_listen_ip`, and `statsd_listen_port` options to `netuitive_configure` @ziggythehamster

# 0.14.0
- added `disk_usage_collector_metrics_whitelist` @ziggythehamster
- improved README.md @ziggythehamster

## 0.13.0
- fix typos in tags and relations @ziggythehamster
- fix typo with logger repo provider @ziggythehamster
- added `docker_collector_enabled` @ziggythehamster


## 0.12.0
- adding some tools for making versioning and tagging easier
- adding doc on chef version compatibility
- restructure testing to allow multiple versions to be developed and tested

## 0.11.0 (cheff 11 branch)
- adding some tools to make versioning easier
- adding doc on chef version compatibility
- chef 11 support (not compatible with 12 so its in its own branch)
- restructure testing to allow multiple versions to be developed and tested

## 0.10.0
- make sure that agent restarts when custom collector config changes (is added).

## 0.9.0
- Update to latest Netuitive-agent (0.2.9-98)

## 0.8.0
- cleanup deprecated code

## 0.7.0
- Update to latest Netuitive-agent (0.2.8-97)

## 0.6.0
- Update to latest Netuitive-agent (0.2.7-96)

## 0.4.0
- Update to latest Netuitive-agent (0.2.6-95)
- Update default netuitive-agent.conf template

## 0.3.1
### Fixed
- bad metadata for platforms supported

## 0.3.0
### Added
- Support for epel based systems

## 0.2.0
### Added
- foodcritic testing
### Changed
- All LWRPs use `use_inline_resources` to address foodcritic ~FC057

## 0.1.1
- Update to latest Netuitive-agent (0.2.3-70)

## 0.1.0
- Initial release of netuitive


libraries/provider_repo.rb
# frozen_string_literal: true

class NetuitiveCookbook::NetuitiveRepoProvider < Chef::Provider::LWRPBase
  include NetuitiveCookbook::Helpers
  provides :netuitive_repo

  use_inline_resources

  action :add do
    platform = determine_platform
    unless new_resource.version
      new_resource.version determine_version(new_resource, node)
    end
    unless new_resource.repo_urls
      new_resource.repo_urls determine_repo_urls(new_resource, node)
    end
    unless new_resource.repo_keys
      new_resource.repo_keys determine_repo_keys(new_resource, node)
    end
    unless new_resource.repo_components
      new_resource.repo_components determine_repo_components(new_resource, node)
    end
    unless new_resource.repo_priority_pins
      new_resource.repo_priority_pins determine_repo_priority_pins(new_resource, node)
    end

    case platform
    when 'debian'
      # add repo and key
      apt_repository 'netuitive' do
        distribution ''
        uri new_resource.repo_urls['debian']
        key new_resource.repo_keys['debian']
        components new_resource.repo_components['debian']
        action :add
      end

      # set apt pin priority
      apt_preference 'netuitive' do
        pin 'version ' + new_resource.version
        pin_priority new_resource.repo_priority_pins['debian']
      end
    when 'rhel', 'fedora', 'amazon'
      Chef::Log.warn 'EPEL based system support is still in the works, please submit an issue on github for any issues you have'
      # we can only use yum-plugin-versionlock if we enable epel repos
      if new_resource.use_epel_repos
        yum_repository 'epel' do
          description 'Extra Packages for Enterprise Linux'
          mirrorlist 'http://mirrors.fedoraproject.org/mirrorlist?repo=epel-6&arch=$basearch'
          gpgkey 'http://dl.fedoraproject.org/pub/epel/RPM-GPG-KEY-EPEL-6'
          action :create
        end
        # add the yum-plugin-versionlock package
        yum_package 'yum-plugin-versionlock' do
          action :install
        end
      else
        Chef::Log.warn 'Warning: We are not going to version lock unless you allow EPEL repos'
      end
      # add the repo
      yum_repository 'netuitive' do
        description 'Netuitive EPEL Repo'
        url new_resource.repo_urls['rhel']
        gpgkey new_resource.repo_keys['rhel']
        gpgcheck true
        priority new_resource.repo_priority_pins['rhel']
      end
    end
  end
end


libraries/provider_collector.rb
# frozen_string_literal: true

class NetuitiveCookbook::NetuitiveCollectorProvider < Chef::Provider::LWRPBase
  include NetuitiveCookbook::Helpers
  provides :netuitive_collector

  use_inline_resources

  action :create do
    unless new_resource.custom_collectors
      new_resource.custom_collectors determine_custom_collectors(new_resource, node)
    end
    unless new_resource.collectors_dir
      new_resource.collectors_dir determine_collectors_dir(new_resource, node)
    end

    # we expect a hash of collectors to gernerate a config for each
    new_resource.custom_collectors.each do |collector, options|
      Chef::Log.warn collector
      options.each do |k, v|
        Chef::Log.warn "#{k}=#{v}"
      end
      template "#{new_resource.collectors_dir}/#{collector}.conf" do
        source 'collector_generic.conf.erb'
        owner 'root'
        group 'root'
        cookbook new_resource.cookbook_template
        mode 0o0744
        variables(
          options: options
        )

        if new_resource.auto_restart_service_on_change
          notifies :restart, 'service[netuitive-agent]', :delayed
        end
      end
    end

    if new_resource.auto_restart_service_on_change
      service 'netuitive-agent' do
        subscribes :restart, new_resource.conf_path.to_s
      end
    end
  end
end


libraries/provider_configure.rb
# frozen_string_literal: true

class NetuitiveCookbook::NetuitiveConfigureProvider < Chef::Provider::LWRPBase
  include NetuitiveCookbook::Helpers
  provides :netuitive_configure

  use_inline_resources

  action :create do
    # your netutive config with the default plugins
    # see custom_collectors to add additional collectors
    template new_resource.conf_path do
      source new_resource.source
      cookbook new_resource.cookbook_template
      variables(
        api_key: new_resource.api_key,
        api_url: new_resource.api_url,
        batch_size: new_resource.batch_size,
        disk_space_collector_exclude_filters: new_resource.disk_space_collector_exclude_filters,
        disk_usage_collector_metrics_whitelist: new_resource.disk_usage_collector_metrics_whitelist,
        docker_collector_enabled: new_resource.docker_collector_enabled,
        docker_collector_metrics_whitelist: new_resource.docker_collector_metrics_whitelist,
        statsd_enabled: new_resource.statsd_enabled,
        statsd_forward: new_resource.statsd_forward,
        statsd_forward_ip: new_resource.statsd_forward_ip,
        statsd_forward_port: new_resource.statsd_forward_port,
        statsd_listen_ip: new_resource.statsd_listen_ip,
        statsd_listen_port: new_resource.statsd_listen_port,
        statsd_prefix: new_resource.statsd_prefix,
        tags: new_resource.tags,
        relations: new_resource.relations
      )

      if new_resource.auto_restart_service_on_change
        notifies :restart, 'service[netuitive-agent]'
      end
    end

    if new_resource.auto_restart_service_on_change
      service 'netuitive-agent' do
        subscribes :restart, new_resource.conf_path.to_s
      end
    end
  end
end


libraries/resource_configure.rb
# frozen_string_literal: true

class NetuitiveCookbook::NetuitiveConfigureResource < Chef::Resource::LWRPBase
  resource_name :netuitive_configure
  provides :netuitive_configure

  actions :create
  default_action :create

  attribute(:api_key, kind_of: String, default: 'CHANGE_ME_PLZ')
  attribute(:api_url, kind_of: String, default: 'https://api.app.netuitive.com/ingest/infrastructure')
  attribute(:auto_restart_service_on_change, kind_of: [TrueClass, FalseClass], default: true)
  attribute(:batch_size, kind_of: Integer, default: 500)
  attribute(:conf_path, kind_of: String, default: '/opt/netuitive-agent/conf/netuitive-agent.conf')
  attribute(:cookbook_template, kind_of: String, default: 'netuitive')
  attribute(:disk_space_collector_exclude_filters, kind_of: [String, Array], default: '^/boot, ^/mnt')
  attribute(:disk_usage_collector_metrics_whitelist, kind_of: String, default: '(?:^.*\.io$|^.*\.average_queue_length$|^.*\.await$|^.*\.iops$|^.*\.read_await$|^.*\.reads$|^.*\.util_percentage|^.*\.write_await$|^.*\.writes$)')
  attribute(:docker_collector_enabled, kind_of: [TrueClass, FalseClass], default: false)
  attribute(:docker_collector_metrics_whitelist, kind_of: String, default: '.*')

  # an array of relations
  attribute(:relations, kind_of: Array, default: [])
  attribute(:source, kind_of: String, default: 'netuitive-agent.conf.erb')

  # statsd attributes
  attribute(:statsd_enabled, kind_of: String, default: 'False')
  attribute(:statsd_forward, kind_of: [TrueClass, FalseClass], default: false)
  attribute(:statsd_forward_ip, kind_of: String, default: '127.0.0.1')
  attribute(:statsd_forward_port, kind_of: Integer, default: 9125)
  attribute(:statsd_listen_ip, kind_of: String, default: '127.0.0.1')
  attribute(:statsd_listen_port, kind_of: Integer, default: 8125)
  attribute(:statsd_prefix, kind_of: String, default: 'statsd')

  # the array needs to be a list of strings with 'tagName:tagVal'
  attribute(:tags, kind_of: Array, default: [])
end
